### PR TITLE
added helper functions and gradient accumulation

### PR DESCRIPTION
**Pull request by brentbiseda**
Added Gradient Accumulation;
Added helper functions for: VGG19, ResNet, Mobile Net -> BaseModel, AttentionModel

### CODE REVIEW

File: src/train.py [hunk 1]
```diff
@@ -267,11 +309,37 @@ def train():
...
         train, labels, params.BATCH_SIZE))
 
-    #model = create_simple_model(labels, sample_X.shape[1:])
-    model = create_VGG19_attention_model(labels, sample_X.shape[1:])
+    adamAccumulate = gradient_accumulation.AdamAccumulate(lr=params.LEARNING_RATE, accum_iters=params.ACCUMULATION_STEPS) #Create Custom Optimizer to use larger synthetic batches
+    
+    '''
+    Pretrained Models from Helper Methods
+    '''
+    #baseModel = _create_MobileNet_model(labels, sample_X.shape[1:], trainable=False, weights="imagenet") # pre-trained requires 224 x 224 input for mobilenet
```
Comment (paloukari) [inline src/train.py]: Can you make this configurable, controlled by either the params or the user input? Code comments are generally not a good practice, normally a code comment means "remove me"

File: src/train.py [hunk 2]
```diff
@@ -267,11 +309,37 @@ def train():
...
 
+    '''
+    Add Attention Layers to Pretrained Models
+    '''
+    #model = create_attention_model(baseModel, labels, optimizer=adamAccumulate)
+    
+    '''
+    Fine Tuning the Model from the Pre-trained Weight Path now with Unfrozen Layers
+    '''
+    #baseModel = _create_InceptionResNetV2_model(labels, sample_X.shape[1:], trainable=True, weights="imagenet")
```
Comment (paloukari) [inline src/train.py]: Same as the above comment

File: src/reset.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+from keras.backend.tensorflow_backend import get_session
+import tensorflow
+import gc
+
+# Reset Keras Session
+def reset_keras():
+    sess = get_session()
+    clear_session()
+    sess.close()
+    sess = get_session()
```
Comment (paloukari) [inline src/reset.py]: Do we need the second `get_session()`?

File: src/reset.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+from keras.backend.tensorflow_backend import clear_session
+from keras.backend.tensorflow_backend import get_session
+import tensorflow
+import gc
+
+# Reset Keras Session
+def reset_keras():
+    sess = get_session()
+    clear_session()
+    sess.close()
```
Comment (paloukari) [inline src/reset.py]: None check here? The Session might have crashed?

File: src/gradient_accumulation.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
+# Gradient Accumulation Borrowed from following location:
+# https://stackoverflow.com/questions/55268762/how-to-accumulate-gradients-for-large-batch-sizes-in-keras
+import keras.backend as K
+from keras.legacy import interfaces
+from keras.optimizers import Optimizer
+
+class AdamAccumulate(Optimizer):
```
Comment (paloukari) [inline src/gradient_accumulation.py:7]: Class and method doc strings would be nice for other people to quickly understand what is this

File: src/reset.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    sess = get_session()
+
+    try:
+        del classifier # this is from global space - change this as you need
```
Comment (paloukari) [inline src/reset.py]: Where exactly is this defined? Can you add a file comment here?

File: src/train.py [hunk 7]
```diff
@@ -235,7 +277,7 @@ def fit_model(model, train, valid):
 
     early = EarlyStopping(monitor="val_loss",
                           mode="min",
-                          patience=5)
+                          patience=params.EARLY_STOPPING)
```
Comment (paloukari) [inline src/train.py:281]: good refactoring, thanks

File: src/train.py [hunk 8]
```diff
@@ -267,11 +309,37 @@ def train():
...
+    #model.load_weights(params.WEIGHT_PATH)
+    
+    '''
+    Train the Model
+    '''
     fit_model(model, train_generator, validation_generator)
 
-
 if __name__ == '__main__':
+    reset.reset_keras() #Garbage Collection
```
Comment (paloukari) [inline src/train.py]: does it make sense to test all networks in the same run? if yes, can you make this be configuration driven (a list of networks to run).

**Review by paloukari — CHANGES_REQUESTED** [review]

File: src/reset.py (hunk 6 above)
Comment (brentbiseda) [reply src/reset.py]: Those lines weren't necessary and were removed

File: src/train.py (hunk 8 above)
Comment (brentbiseda) [reply src/train.py]: Added helper function to train all 3 simple models for comparison.  We can then run training on the selected attention model

File: src/gradient_accumulation.py (hunk 5 above)
Comment (brentbiseda) [reply src/gradient_accumulation.py:7]: Added high level doc string

File: src/reset.py (hunk 3 above)
Comment (brentbiseda) [reply src/reset.py]: Should not be necessary.  Removed

File: src/train.py (hunk 1 above)
Comment (brentbiseda) [reply src/train.py]: Added print statement and reverts to random initialization if not appropriate size.  This issue is only specific to mobile net in Keras.

**Comment by brentbiseda** [comment]
See my changes

**Review by paloukari — APPROVED** [review]
